File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int rank = 0;
    int val = 0;
    public int kthSmallest(TreeNode root, int k) {
        inorder(root, k);
        return val;
    }
    public void inorder(TreeNode root, int k) {
        if(root == null) return;
        inorder(root.left, k);
        rank++;
        if(rank == k) {
            val = root.val;
            return;
        }
        inorder(root.right, k);
    }
}

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        cnt = 0
        res = 0
        def inorder(root):
            nonlocal cnt, res
            if not root:
                return
            inorder(root.left)
            cnt += 1
            if cnt == k:
                res= root.val
            inorder(root.right)
        inorder(root)
        return res